### DISCUSSION TITLE
[Enhancement] [carousel] 暴露 `getActiveIndex` 函数

### DISCUSSION THREAD

**Discussion by a145789** [element-plus/element-plus#17648] [Feature Request]
Feature type: **`Enhancement`**

## Details

- Enhanced component name: `carousel`

## Description

`carousel`组件的 `change` 事件会在当前 nextTick 后 [触发](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtY2Fyb3VzZWwgaW5kaWNhdG9yLXBvc2l0aW9uPVwibm9uZVwiIEBjaGFuZ2U9XCJoYW5kbGVDaGFuZ2VcIiA6YXV0b3BsYXk9XCJmYWxzZVwiIHR5cGU9XCJjYXJkXCIgcmVmPVwiY2Fyb3VzZWxSZWZcIj5cbiAgICA8ZWwtY2Fyb3VzZWwtaXRlbSB2LWZvcj1cIml0ZW0gaW4gNFwiIDprZXk9XCJpdGVtXCI+XG4gICAgICA8aDMgQGNsaWNrPVwiY2xpY2tIM1wiPlxuICAgICAgICB7eyBpdGVtIH19XG4gICAgICA8L2gzPlxuICAgIDwvZWwtY2Fyb3VzZWwtaXRlbT5cbiAgPC9lbC1jYXJvdXNlbD5cbjwvdGVtcGxhdGU+XG5cbjxzY3JpcHQgbGFuZz1cInRzXCIgc2V0dXA+XG5pbXBvcnQgeyBFbE1lc3NhZ2UgfSBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBuZXh0VGljaywgd2F0Y2gsIHJlZiB9IGZyb20gJ3Z1ZSc7XG5cbmNvbnN0IGNhcm91c2VsUmVmID0gcmVmKG51bGwpXG5cbmNvbnN0IGluZGV4MSA9IHJlZignMCcpXG5mdW5jdGlvbiBoYW5kbGVDaGFuZ2UoaW5kZXg6IHN0cmluZykge1xuICBjb25zb2xlLmxvZygnY2hhbmdlJywgaW5kZXgpXG4gIGluZGV4MS52YWx1ZSA9IGluZGV4XG59XG5cbndhdGNoKGluZGV4MSwgKCkgPT4ge1xuICBjb25zb2xlLmxvZygnd2F0Y2gnLCBpbmRleDEudmFsdWUpXG59LCB7XG4gIGltbWVkaWF0ZTogdHJ1ZVxufSlcblxuZnVuY3Rpb24gY2xpY2tIMygpIHtcbiAgbmV4dFRpY2soKCkgPT4ge1xuICAgIGNvbnNvbGUubG9nKCdjbGljaycsIGluZGV4MS52YWx1ZSlcbiAgfSlcbn1cbjwvc2NyaXB0PlxuXG48c3R5bGUgc2NvcGVkPlxuLmVsLWNhcm91c2VsX19pdGVtIGgzIHtcbiAgY29sb3I6ICM0NzU2Njk7XG4gIGZvbnQtc2l6ZTogMThweDtcbiAgb3BhY2l0eTogMC43NTtcbiAgbGluZS1oZWlnaHQ6IDMwMHB4O1xuICBtYXJnaW46IDA7XG4gIHRleHQtYWxpZ246IGNlbnRlcjtcbn1cblxuLmVsLWNhcm91c2VsX19pdGVtOm50aC1jaGlsZCgybikge1xuICBiYWNrZ3JvdW5kLWNvbG9yOiAjOTlhOWJmO1xufVxuXG4uZWwtY2Fyb3VzZWxfX2l0ZW06bnRoLWNoaWxkKDJuICsgMSkge1xuICBiYWNrZ3JvdW5kLWNvbG9yOiAjZDNkY2U2O1xufVxuPC9zdHlsZT5cbiIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHtcbiAgICBcInZ1ZVwiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vQHZ1ZS9ydW50aW1lLWRvbUBsYXRlc3QvZGlzdC9ydW50aW1lLWRvbS5lc20tYnJvd3Nlci5qc1wiLFxuICAgIFwiQHZ1ZS9zaGFyZWRcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvc2hhcmVkQGxhdGVzdC9kaXN0L3NoYXJlZC5lc20tYnVuZGxlci5qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9lbGVtZW50LXBsdXNAbGF0ZXN0L2Rpc3QvaW5kZXguZnVsbC5taW4ubWpzXCIsXG4gICAgXCJlbGVtZW50LXBsdXMvXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9lbGVtZW50LXBsdXNAbGF0ZXN0L1wiLFxuICAgIFwiQGVsZW1lbnQtcGx1cy9pY29ucy12dWVcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlQDIvZGlzdC9pbmRleC5taW4uanNcIlxuICB9LFxuICBcInNjb3Blc1wiOiB7fVxufSIsImVsZW1lbnQtcGx1cy5qcyI6ImltcG9ydCB7IGdldEN1cnJlbnRJbnN0YW5jZSB9IGZyb20gJ3Z1ZSdcbmltcG9ydCBFbGVtZW50UGx1cyBmcm9tICdlbGVtZW50LXBsdXMnXG5cbmxldCBpbnN0YWxsZWQgPSBmYWxzZVxuYXdhaXQgbG9hZFN0eWxlKClcblxuZXhwb3J0IGZ1bmN0aW9uIHNldHVwRWxlbWVudFBsdXMoKSB7XG4gIGlmIChpbnN0YWxsZWQpIHJldHVyblxuICBjb25zdCBpbnN0YW5jZSA9IGdldEN1cnJlbnRJbnN0YW5jZSgpXG4gIGluc3RhbmNlLmFwcENvbnRleHQuYXBwLnVzZShFbGVtZW50UGx1cylcbiAgaW5zdGFsbGVkID0gdHJ1ZVxufVxuXG5leHBvcnQgZnVuY3Rpb24gbG9hZFN0eWxlKCkge1xuICBjb25zdCBzdHlsZXMgPSBbJ2h0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vZWxlbWVudC1wbHVzQGxhdGVzdC9kaXN0L2luZGV4LmNzcycsICdodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL2VsZW1lbnQtcGx1c0BsYXRlc3QvdGhlbWUtY2hhbGsvZGFyay9jc3MtdmFycy5jc3MnXS5tYXAoKHN0eWxlKSA9PiB7XG4gICAgcmV0dXJuIG5ldyBQcm9taXNlKChyZXNvbHZlLCByZWplY3QpID0+IHtcbiAgICAgIGNvbnN0IGxpbmsgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCdsaW5rJylcbiAgICAgIGxpbmsucmVsID0gJ3N0eWxlc2hlZXQnXG4gICAgICBsaW5rLmhyZWYgPSBzdHlsZVxuICAgICAgbGluay5hZGRFdmVudExpc3RlbmVyKCdsb2FkJywgcmVzb2x2ZSlcbiAgICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignZXJyb3InLCByZWplY3QpXG4gICAgICBkb2N1bWVudC5ib2R5LmFwcGVuZChsaW5rKVxuICAgIH0pXG4gIH0pXG4gIHJldHVybiBQcm9taXNlLmFsbChzdHlsZXMpXG59XG4iLCJ0c2NvbmZpZy5qc29uIjoie1xuICBcImNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJ0YXJnZXRcIjogXCJFU05leHRcIixcbiAgICBcImpzeFwiOiBcInByZXNlcnZlXCIsXG4gICAgXCJtb2R1bGVcIjogXCJFU05leHRcIixcbiAgICBcIm1vZHVsZVJlc29sdXRpb25cIjogXCJCdW5kbGVyXCIsXG4gICAgXCJ0eXBlc1wiOiBbXCJlbGVtZW50LXBsdXMvZ2xvYmFsLmQudHNcIl0sXG4gICAgXCJhbGxvd0ltcG9ydGluZ1RzRXh0ZW5zaW9uc1wiOiB0cnVlLFxuICAgIFwiYWxsb3dKc1wiOiB0cnVlLFxuICAgIFwiY2hlY2tKc1wiOiB0cnVlXG4gIH0sXG4gIFwidnVlQ29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiAzLjNcbiAgfVxufVxuIiwiUGxheWdyb3VuZE1haW4udnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCBBcHAgZnJvbSAnLi9BcHAudnVlJ1xuaW1wb3J0IHsgc2V0dXBFbGVtZW50UGx1cyB9IGZyb20gJy4vZWxlbWVudC1wbHVzLmpzJ1xuc2V0dXBFbGVtZW50UGx1cygpXG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8QXBwIC8+XG48L3RlbXBsYXRlPlxuIiwiX28iOnt9fQ==)，这对 nextTick 很不友好，暴露 `getActiveIndex` 方便外部可以在 nextTick 中获取到最新的 activeIndex

### Additional comments

_(empty)_

**Comment by warmthsea** [comment]
https://github.com/element-plus/element-plus/pull/17650

**Comment by tolking** [comment]
因为这样的执行顺序，暴露了 getActiveIndex 也会拿不到最新值。建议先优化下代码使用

**Reply by a145789** [reply]
> 因为这样的执行顺序，暴露了 getActiveIndex 也会拿不到最新值。建议先优化下代码使用

看了下 element 的源码, `activeIndex` 只能通过 `setActiveItem` 函数修改，`setActiveItem`  函数内部没有异步， `activeIndex` 是同步修改的，在 element 内部 [源码](https://github.com/element-plus/element-plus/blob/a11a7d09dc53ab15dc8580a80cc49ddcf0ee58db/packages/components/carousel/src/use-carousel-item.ts#L130~L137)，也只有此处同步调用 `setActiveItem` ，只要触发了 `handleItemClick` ，不管是哪里的 nextTick 回调肯定可以获取到最新 activeIndex。

而只针对我的场景的话，更合理的需求是再加一个 `onItemClick` 事件，我想要在 item click 后做一些事。

暴露 getActiveIndex 在一些场景很实用，比如，在用户点击保存后向后端发送当前轮播图的index，不需要其他额外操作，现在只能自己定义一个变量用 change 事件同步内部的 activeIndex，感觉直接用   getActiveIndex 可能会更好点？

**Reply by tolking** [reply]
真实的执行顺序和你想的可能不一样，[一个最基础的示例](https://play.vuejs.org/#__SSR__eNp9UstOwzAQ/JWVLwQRpRQ4oRTxEIdyAAQcfYncbWNw7Mh2QqUo/87aUWhaVT1EindnxzOz7thDXWdtg+yW5U5YWXtw6Jv6jmtZ1cZ66OC38KJMQePWf0nxk4LFNfSwtqaCM5o945prYbTzUAgvW1zqFW5hEXDJ5fmuWxZ6pXDpsXpSRESI5BwWd9BxDdPRrC1UgxcXoRwmjcJMmU1yRVz9Id0RqlFpMqntM82JCaAf+aLDZKIgndJNB6/DCH35bEiLcqIDOapV4ZFOAPlKtnAvgqwFZweeOYsYQpXzQ9AIGBx13V6afZ/PyvlwwYxuoL98NrmXpcw7UrqWm+zbGU0LjeI5E6aqpUL7VntJTji7HRPhrFDK/L7EmrcNpmNdlCh+jtS/3TbUOHu36NC2yNl/zxd2g35oP3++0g4mzcqsGkXoE80PpJSboHGAPTYhFDvBRbXL+Cyl3ny5561H7UZTQWjcasRzRi/z6YT1ndzr7CbO0WZZ/wcmgAVE)

导出 activeIndex 看上去是个合理的需求。Link #17650
